### PR TITLE
Use bash on windows

### PR DESCRIPTION
**Pull request by alimcmaster1**
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- towards https://github.com/pandas-dev/pandas/issues/26344

As per travis logs:
https://travis-ci.org/pandas-dev/pandas/jobs/552444580

```
ccache: /usr/bin/ccache
source deactivate
DeprecationWarning: 'source deactivate' is deprecated. Use 'conda deactivate'.
```
As of [conda 4.6 the preferred](https://www.anaconda.com/conda-4-6-release/) way of using an environment is to use `conda activate (env)`.
`source activate` will eventually be deprecated 

(only remaining place to do this is in `code_checks.yml` -> I will do this in a follow up for ease of review) 

* This make the steps in the `windows.yml` and `posix.yml` file very similar and we can potentially merge these.

* Use `run_tests.sh` on windows this will mean  we now tests with `pytest -n 2` and `pytest -n 1` on windows - now consistent with the linux builds

cc. @TomAugspurger @datapythonista

### CODE REVIEW

**Comment by jreback** [comment]
sure

**Comment by jorisvandenbossche** [comment]
@alimcmaster1 I was trying a few days ago something similar for geopandas (although for appveyor), and needed some PATH manipulations in the end to get this working (with the help of @msarahan) , see https://github.com/geopandas/geopandas/pull/1029. It might be that something similar is needed here, if the `conda init bash` does not work on CI.

**Comment by alimcmaster1** [comment]
Awesome thanks for the pointer @jorisvandenbossche

**Comment by jbrockmendel** [comment]
@jorisvandenbossche is this close to being ready?

**Comment by alimcmaster1** [comment]
@jbrockmendel - still needs a bit of work. CI is currently only passing on windows. I will take another look

**Comment by alimcmaster1** [comment]
Thanks @WillAyd will give that a try

File: ci/azure/posix.yml [hunk 1]
```diff
@@ -56,15 +56,20 @@ jobs:
   steps:
     - script: |
         if [ "$(uname)" == "Linux" ]; then sudo apt-get install -y libc6-dev-i386 $EXTRA_APT; fi
-        echo '##vso[task.prependpath]$(HOME)/miniconda3/bin'
+        echo '##vso[task.setvariable variable=MINICONDA_DIR]$(HOME)/miniconda3'
```
Comment (WillAyd) [inline ci/azure/posix.yml]: I think this needs to be executed by `bash` and not `script` (which on windows is just cmd.exe). Not sure if there is an overarching reason for this and similar blocks to be `script` but as mentioned changing to `bash` might fix the issue remaining

File: ci/azure/windows.yml [hunk 2]
```diff
@@ -20,20 +20,23 @@ jobs:
...
+      displayName: 'Update conda'
+    - bash: |
+        source $HOME/miniconda3/etc/profile.d/conda.sh
+        conda activate pandas-dev
         conda env create -q --file ci\\deps\\azure-windows-$(CONDA_PY).yaml
       displayName: 'Create anaconda environment'
     - script: |
-        call activate pandas-dev
-        call conda list
+        call conda.bat activate pandas-dev
```
Comment (WillAyd) [inline ci/azure/windows.yml]: Why not just change all of the `script` commands to `bash` and then just use without `call`? Right now the usage looks mixed so would make things consistent and easier to read I think for non-Windows people

File: ci/azure/windows.yml [hunk 3]
```diff
@@ -18,22 +18,26 @@ jobs:
 
   steps:
     - powershell: |
-        Write-Host "##vso[task.prependpath]$env:CONDA\Scripts"
-        Write-Host "##vso[task.prependpath]$HOME/miniconda3/bin"
```
Comment (alimcmaster1) [inline ci/azure/windows.yml]: Correct this to miniconda dir which is $env:CONDA

File: ci/azure/windows.yml [hunk 4]
```diff
@@ -18,22 +18,27 @@ jobs:
...
+        Write-Host "##vso[task.prependpath]$env:CONDA\condabin"
+        Write-Host "##vso[task.setvariable variable=MINICONDA_DIR]$env:CONDA"
       displayName: 'Add conda to PATH'
-    - script: conda update -q -n base conda
-      displayName: Update conda
-    - script: |
-        call activate
-        conda env create -q --file ci\\deps\\azure-windows-$(CONDA_PY).yaml
+    - bash: |
+        source $MINICONDA_DIR/etc/profile.d/conda.sh
```
Comment (WillAyd) [inline ci/azure/windows.yml]: Is this necessary with the vso task above?

File: ci/azure/windows.yml [hunk 5]
```diff
@@ -18,22 +18,27 @@ jobs:
...
+        conda update -q -y -n base conda
+      displayName: 'Update conda'
+    - bash: |
+        source $MINICONDA_DIR/etc/profile.d/conda.sh
```
Comment (WillAyd) [inline ci/azure/windows.yml]: Same question here; I think the task already puts conda on the path so no need to run this script (assuming that's what it does)

File: ci/azure/windows.yml (hunk 2 above)
Comment (alimcmaster1) [reply ci/azure/windows.yml]: changed all of `windows.yml` file `script` steps to use `bash` instead

File: ci/azure/windows.yml (hunk 5 above)
Comment (alimcmaster1) [reply ci/azure/windows.yml]: i found this was necessary too - otherwise conda complains you need to run `conda init`. Let me double check this though.  sourcing conda.sh here seemed to handle everything nicely

**Comment by alimcmaster1** [comment]
This is almost over the line - I will update the windows CI to use run_tests.sh so its consistent with Linux CI

File: ci/azure/windows.yml (hunk 5 above)
Comment (alimcmaster1) [reply ci/azure/windows.yml]: confirmed I definitely need this otherwise we get:
```
CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.
```

**Comment by alimcmaster1** [comment]
@WillAyd i've fixed this up + edited description highlighting what this hopes to achieve mind taking a look?

File: ci/azure/windows.yml [hunk 6]
```diff
@@ -11,49 +11,69 @@ jobs:
       py36_np15:
         ENV_FILE: ci/deps/azure-windows-36.yaml
         CONDA_PY: "36"
+        PATTERN: "not slow and not network"
 
       py37_np141:
         ENV_FILE: ci/deps/azure-windows-37.yaml
         CONDA_PY: "37"
+        PATTERN: "not slow and not network"
```
Comment (datapythonista) [inline ci/azure/windows.yml]: Is this related? If not, I think we should have it in a separate PR.

**Review by datapythonista — CHANGES_REQUESTED** [review]
I'm a bit unsure on this... I think changing `source deactivate` by `conda deactivate` is trivial, and fixes the warning.

Didn't check the windows build in detail, but I don't think in Linux there is any warning on using `source activate`, and we are significantly increasing the complexity of the build to make it a `conda activate`.

I'd open separate PRs, as there are surely good changes here. But replacing `source activate pandas-dev` by `source $MINICONDA_DIR/etc/profile.d/conda.sh && conda activate pandas-dev` doesn't seem reasonable to me.

File: ci/azure/windows.yml (hunk 6 above)
Comment (alimcmaster1) [reply ci/azure/windows.yml]: Okay agree - let me factor this out to a separate PR. This was part of the work to move the windows CI scripts to use bash.  (FYI I'm away from my keyboard for a few weeks - but will work on this when I return)

**Comment by WillAyd** [comment]
> But replacing `source activate pandas-dev` by `source $MINICONDA_DIR/etc/profile.d/conda.sh && conda activate pandas-dev` doesn't seem reasonable to me.

Yea agreed on that. So doesn't using `bash` instead of `script` get rid of the need for the former? I thought the point of the VSO task was to enable simple `conda` usage without `source` so long as `bash` is used. I'm also not sure there's a point to using `script` / `cmd.exe` on Windows

**Comment by alimcmaster1** [comment]
Going to wait till https://github.com/pandas-dev/pandas/pull/28531 is merged then I will open a separate PR to update `windows.yml` to use bash in the steps opposed to script. As per #26344. Unless any dev's disagree?

**Comment by datapythonista** [comment]
Sounds good  thanks @alimcmaster1

File: .travis.yml [hunk 7]
```diff
@@ -88,11 +88,12 @@ before_script:
...
   - ci/run_tests.sh
 
 after_script:
   - echo "after_script start"
-  - source activate pandas-dev && pushd /tmp && python -c "import pandas; pandas.show_versions();" && popd
-  - ci/print_skipped.py 
-  - echo "after_script done"
+  - conda activate pandas-dev && pushd /tmp && python -c "import pandas; pandas.show_versions();" && popd
+  - ci/print_skipped.py
+  - echo "after_script done"
```
Comment (jreback) [inline .travis.yml]: need a newline

File: ci/azure/posix.yml [hunk 8]
```diff
@@ -95,6 +100,7 @@ jobs:
         }
       displayName: 'Check for test failures'
     - script: |
-        source activate pandas-dev
+        source $MINICONDA_DIR/etc/profile.d/conda.sh
+        conda activate pandas-dev
         python ci/print_skipped.py
-      displayName: 'Print skipped tests'
+      displayName: 'Print skipped tests'
```
Comment (jreback) [inline ci/azure/posix.yml]: add a newline

File: ci/azure/windows.yml (hunk 6 above)
Comment (jreback) [reply ci/azure/windows.yml]: can you factor this out

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
can you merge master

**Comment by alimcmaster1** [comment]
Will do - waiting for @datapythonista to merge his PR first

**Comment by jbrockmendel** [comment]
> waiting for datapythonista to merge his PR first

can you rebase here anyway?

**Comment by jbrockmendel** [comment]
@alimcmaster1 can you pls rebase?

**Comment by jbrockmendel** [comment]
@alimcmaster1 it looks like you just pushed, but GH still says there are conflicts

**Comment by alimcmaster1** [comment]
@jbrockmendel sorry for the delay updated. We should wait till https://github.com/pandas-dev/pandas/pull/28531 is merged as in my PR here we are currently calling pytest with n = 2 and n=1 on windows where as the current behaviour just calls:
`pytest -m "not slow and not network" --junitxml=test-data.xml pandas -n 2 -r sxX --strict --durations=10 %*`

This change means the steps in both `posix.yml` and `windows.yml` should be similar. (and we can potentially merge)

**Comment by alimcmaster1** [comment]
Closing as the scope of this changed slight. Will summarise and open fresh PR - to make this easier to review.

Thanks all for the comments

**Comment by datapythonista** [comment]
Also worth considering that we can move things to GitHub Actions
